Properly include generated Python files in the wheel

llucax opened the pull request:
When building the wheel, the setuptools file discovery feature runs before the `compile_betterproto` sub-command is run, so the files are not included in the resulting wheel.

To overcome this issue we use a hack, by running the compile command early when the distribution options are being finalized, we make sure the files are present by the time the file discovery runs.

This is very hacky though, and even when it seems to work well in all tested cases (calling setuptools` `setup()` directly, `build`, `pip install`, `pip install -e`), I wouldn't be very surprised if it breaks in the future, so we probably will have to have an eye on it.

Unfortunately it seems there is no easy way to do this with setuptools:

* https://github.com/pypa/setuptools/discussions/3180#discussioncomment-2422328
* https://github.com/pypa/setuptools/issues/2591

This PR also includes some other improvements, including:

* Better abstraction of the configuration object
* Better hooking into setuptools
* Use of logging instead of print
* Better name for the command to compile the proto files

llucax committed:
Rename `build_betterproto` to `compile_betterproto`

The protobuf files are said to be "compiled", so it is better to stick
to that term to avoid confusion.

Also rename the class `CompileProto` to `CompileBetterproto` for more
consistency.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Add a `from_strings()` method to `ProtobufConfig`

This is to easily load the configuration from the plain strings provided
by setuptools commands. The `ProtobufConfig` class will be expanded to
handle all operations related to the configuration.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Store the config object in the command

This is to make it easier to access to the config in a more convenient
way, and to keep moving configuration related processing to the config
object.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Move the expansion of the paths to `ProtobufConfig`

This simplifies the command a lot and let us reuse that expansion in
the future.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Add a `BaseProtoCommand`

This will be used as the base class for other commands that need to
read the protobuf configuration.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Add `@override` to override methods

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Use `logging` instead of `print()`

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Install the `compile_betterproto` command more cleanly

Instead of hijacking the `build` class by importing it in our command
we use a different entry point,
`setuptools.finalize_distribution_options`, to add our build
sub-command.

This is still not a 100% clean, as that entry point is not specifically
for this purpose, but it is still cleaner than the previous approach.

For more details, see:

* https://github.com/pypa/setuptools/discussions/3180#discussioncomment-2422328
* https://setuptools.pypa.io/en/latest/userguide/extension.html#customizing-distribution-options
* https://github.com/abravalheri/experiment-setuptools-plugin

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Distribute protobuf files automatically

We now add the protobuf files automatically to the source distribution,
so it is not necessary to manually add them via the `MANIFEST.in` file.

To do this a new `AddProtoFiles` command is created, which is invoqued
by a custom `sdist` command (implemented by `SdistWithProtoFiles`).

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Properly include generated Python files in the wheel

When building the wheel, the setuptools file discovery feature runs
before the `compile_betterproto` sub-command is run, so the files are
not included in the resulting wheel.

To overcome this issue we use a hack, by running the compile command
early when the distribution options are being finalized, we make sure
the files are present by the time the file discovery runs.

This is very hacky though, and even when it seems to work well in all
tested cases (calling setuptools` `setup()` directly, `build`,
`pip install`, `pip install -e`), I wouldn't be very surprised if it
breaks in the future, so we probably will have to have an eye on it.

Unfortunately it seems there is no easy way to do this with setuptools:

* https://github.com/pypa/setuptools/discussions/3180#discussioncomment-2422328
* https://github.com/pypa/setuptools/issues/2591

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Update release notes

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

daniel-zullo-frequenz commented on src/setuptools_betterproto/_config.py:
```diff
@@ -99,3 +100,13 @@ def from_strings(
+        return [
+            str(proto_file.relative_to(proto_path))
+            for proto_file in proto_path.rglob(self.proto_glob)
+        ]
```
there is double return here but it seems it was fixed in another commit

daniel-zullo-frequenz replied:
it was fixed in the commit "Distribute protobuf files automatically"

llucax replied:
Ah, thanks, there was a lot of rebasing and cherry-picking of diff lines, I guess I messed it up at some point, will fix!

daniel-zullo-frequenz commented on src/setuptools_betterproto/_command.py:
```diff
@@ -59,53 +67,99 @@ class CompileProto(setuptools.Command):
+
+    def copy_with_path(self, src: str, dest: str) -> None:
+        """Copy a file from src to dest, creating the destination directory."""
+        if not os.path.exists(os.path.dirname(dest)):
+            _logger.debug("creating directory %s", os.path.dirname(dest))
+            os.makedirs(os.path.dirname(dest))
```
optionally, you might want to get the destination directory only once:
```suggestion
        dest_dir = os.path.dirname(dest)
        if not os.path.exists(dest_dir):
            _logger.debug("creating directory %s", dest_dir)
            os.makedirs(dest_dir)
```

llucax replied:
Yeah, good idea. I also updated the function name and improved the docs a bit.

daniel-zullo-frequenz commented on src/setuptools_betterproto/_command.py:
```diff
@@ -59,53 +67,99 @@ class CompileProto(setuptools.Command):
+
+        if include_files and not proto_files:
+            _logger.warning(
+                "Some proto files (%s) were found in the `include_paths` (%s), but "
```
what do you think about?
```suggestion
                "Some include files (%s) were found in the `include_paths` (%s), but "
```

llucax replied:
I think the current wording is more clear. We found proto files, they are just in the `include_paths` instead of the `proto_path`, but they are still proto files IMHO.

daniel-zullo-frequenz reviewed (commented):
I have a few minor comments to check for, and LGTM

Marenz commented on src/setuptools_betterproto/_command.py:
```diff
@@ -113,3 +116,50 @@ def run(self) -> None:
+            _logger.warning(
+                "Some proto files (%s) were found in the `include_paths` (%s), but "
+                "no proto files where found in the `proto_path` (%s). We are not adding "
+                "the found include files to the source distribution automatically!",
```
How _would_ we add them? Maybe a hint in the warning?

llucax replied:
You mean how would you add the proto files in the `include_paths` manually? I think you don't want to do that, you probably just forgot to configure or misconfigured the `proto_path`. Maybe I can add this to the warning.

Marenz commented on src/setuptools_betterproto/_command.py:
```diff
@@ -113,3 +116,50 @@ def run(self) -> None:
+                self.config.proto_path,
+            )
+
+        if not proto_files:
```
Is this case common or does it denote a problem?

llucax replied:
It would be weird that you use this setuptools plugin and you don't have proto files, ~but a warning will be shown when compiling the files at a later stage. This is just a hack to overcome the issue with the file discovery. I will add a comment to clarify.~ I thought this comment was in a different code path, if you are only building the source distribution, then you won't get the warning because the proto files will not be compiled, so I will add a warning here too.

Marenz commented:
Approved, but removed from the queue in case you want to react to the feedback

llucax committed:
Improve warning message when no proto files are found

When no proto files are found in the `proto_path` using the `proto_glob`,
we show the value of those two configuration options in the warning
message and suggest the user to check if they are configured correctly.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax commented:
Updated:

* Improved the warnings to add more hints on the impact and how to fix them and added a missing warning
* Calculated the `dest_dir` only once
* Renamed `copy_with_path` to `copy_with_directories` and improved the docs
* Fixed the misplaced `return` in the commits

llucax commented:
Re-enabling auto-merge.

daniel-zullo-frequenz commented on src/setuptools_betterproto/_command.py:
```diff
@@ -59,53 +67,120 @@ class CompileProto(setuptools.Command):
+        if include_files and not proto_files:
+            _logger.warning(
+                "Some proto files (%s) were found in the `include_paths` (%s), but "
+                "no proto files where found in the `proto_path`. You probably want to "
```
```suggestion
                "no proto files were found in the `proto_path`. You probably want to "
```

llucax replied:
Fixed!

daniel-zullo-frequenz reviewed (commented):
Just a typo, LGTM and I'm happy to approve as it is. Just let me know
